File: app/src/design/java/dolphin/android/apps/CpblCalendar/provider/TeamHelper.java
package dolphin.android.apps.CpblCalendar.provider;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Color;
import android.support.v7.graphics.Palette;

import dolphin.android.apps.CpblCalendar.CpblApplication;

/**
 * Created by jimmyhu on 2017/3/17.
 * <p>
 * A helper class to handle Team logo & palette.
 */

public class TeamHelper {
    private final CpblApplication mApplication;

    public TeamHelper(CpblApplication application) {
        mApplication = application;
    }

    private CpblApplication getApplication() {
        return mApplication;
    }

    public int getLogoColorFilter(Team team, int year) {
        int logoId = Team.getTeamLogo(team.getId(), year);
        Palette palette = getApplication().getTeamLogoPalette(logoId);
        if (palette == null) {
            Bitmap bitmap = BitmapFactory.decodeResource(getApplication().getResources(), logoId);
            palette = Palette.from(bitmap).generate();
            getApplication().setTeamLogoPalette(logoId, palette);//save to cache
        }
        int color = Color.BLACK;
        if (!validateColor(color) && palette.getDominantSwatch() != null) {
            color = palette.getDominantColor(color);
        }
        if (!validateColor(color) && palette.getVibrantSwatch() != null) {
            color = palette.getVibrantColor(color);
        }
        if (!validateColor(color) && palette.getMutedSwatch() != null) {
            color = palette.getMutedColor(color);
        }
        if (!validateColor(color)) {
            color = Color.BLACK;//still use black
        }
        return color;
    }

    private boolean validateColor(int color) {
        switch (color) {
            case Color.BLACK:
            case Color.WHITE:
            case Color.TRANSPARENT:
                return false;
        }
        return true;
    }
}

File: app/src/design/java/dolphin/android/apps/CpblCalendar/provider/TeamHelper.kt
package dolphin.android.apps.CpblCalendar.provider

import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Color
import androidx.palette.graphics.Palette

import dolphin.android.apps.CpblCalendar3.CpblApplication

/**
 * Created by jimmyhu on 2017/3/17.
 *
 *
 * A helper class to handle Team logo & palette.
 */

class TeamHelper(private val application: CpblApplication) {

    fun getLogoColorFilter(team: Team, year: Int): Int {
        val logoId = Team.getTeamLogo(team.id, year)
        var palette: Palette? = application.getTeamLogoPalette(logoId)
        if (palette == null) {
            val bitmap = BitmapFactory.decodeResource(application.resources, logoId)
            palette = Palette.from(bitmap).generate()
            application.setTeamLogoPalette(logoId, palette)//save to cache
        }
        var color = Color.BLACK
        if (!validateColor(color) && palette.dominantSwatch != null) {
            color = palette.getDominantColor(color)
        }
        if (!validateColor(color) && palette.vibrantSwatch != null) {
            color = palette.getVibrantColor(color)
        }
        if (!validateColor(color) && palette.mutedSwatch != null) {
            color = palette.getMutedColor(color)
        }
        if (!validateColor(color)) {
            color = Color.BLACK//still use black
        }
        return color
    }

    private fun validateColor(color: Int): Boolean {
        when (color) {
            Color.BLACK, Color.WHITE, Color.TRANSPARENT -> return false
        }
        return true
    }
}
